src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs

```diff
@@ -128,31 +159,122 @@ public IEnumerable<ReplicationBatchItem> GetTombstonesByBucketFrom(DocumentsOper
+        long numOfDeletions = 0;
+
+        MarkTombstonesAsArtificial(context, bucket);
+
+        foreach (var document in GetDocumentsByBucketFrom(context, bucket, 0))
         {
-            return table.DeleteForwardFrom(dynamicIndex, prefix, startsWith: true, MaxDocumentsToDeleteInBucket, shouldAbort: holder =>
-            {
-                var changeVector = TableValueToChangeVector(context, changeVectorPosition, ref holder.Reader);
-                var status = ChangeVectorUtils.GetConflictStatus(changeVector, upTo);
-                return status != ConflictStatus.AlreadyMerged;
-            });
+            if (numOfDeletions > MaxDocumentsToDeleteInBucket)
+                break;
+
+            var docCv = context.GetChangeVector(document.ChangeVector);
+            if (ChangeVectorUtils.GetConflictStatus(docCv, upTo) != ConflictStatus.AlreadyMerged)
+                break;
+
+            // check change vectors of all document extensions
+            if (HasDocumentExtensionWithGreaterChangeVector(context, document.LowerId, upTo)) 
+                break;
+            
+            Delete(context, document.Id, document.Flags | DocumentFlags.Artificial);
+
+            // delete revisions for document
+            RevisionsStorage.DeleteRevisionsFor(context, document.Id, artificial: true);
+
+            numOfDeletions++;
         }
+
+        return numOfDeletions;
     }
 
-    public static long DeleteItemsByBucketForDocuments(DocumentsOperationContext context, CollectionName collection, int bucket, ChangeVector upTo)
+    private void MarkTombstonesAsArtificial(DocumentsOperationContext context, int bucket)
     {
-        var table = context.Transaction.InnerTransaction.OpenTable(DocsSchema, collection.GetTableName(CollectionTableType.Documents));
-        return DeleteItemsByBucket(context, table, DocsSchema.DynamicKeyIndexes[CollectionDocsBucketAndEtagSlice], (int)DocumentsTable.ChangeVector, bucket, upTo);
+        var readTable = new Table(TombstonesSchema, context.Transaction.InnerTransaction);
+
+        foreach (var result in GetItemsByBucket(context.Allocator, readTable, TombstonesSchema.DynamicKeyIndexes[TombstonesBucketAndEtagSlice], bucket, etag: 0))
+        {
+            var tombstone = TableValueToTombstone(context, ref result.Result.Reader);
+            if (tombstone.Flags.Contain(DocumentFlags.Artificial))
+                continue;
+
+            var collection = TableValueToId(context, (int)TombstoneTable.Collection, ref result.Result.Reader);
+            if (_collectionsCache.TryGetValue(collection, out var collectionName) == false)
+            {
+                collectionName = new CollectionName(collection);
+                _collectionsCache.TryAdd(collection, collectionName);
+            }
+            
+            var writeTable = context.Transaction.InnerTransaction.OpenTable(TombstonesSchema, collectionName.GetTableName(CollectionTableType.Tombstones));
```

Comment:
can we do that? iterate and modify? cc @arekpalinski

Reply:
Once we do a write to a tree then the tree iterator might not be valid any longer. IIRC In other places we open the iterator again after doing a modification to a tree.

Reply:
done